pkg/util/push/push.go

```diff
@@ -121,15 +137,56 @@ func handler(maxRecvMsgSize int,
 				level.Warn(logger).Log("msg", "push request canceled", "err", err)
 				return
 			}
+			var (
+				code int
+				msg  string
+			)
 			resp, ok := httpgrpc.HTTPResponseFromError(err)
-			if !ok {
-				http.Error(w, err.Error(), http.StatusInternalServerError)
-				return
+			if ok {
+				code, msg = int(resp.Code), string(resp.Body)
+			} else {
+				code, msg = getHTTPStatusAndMessage(err)
 			}
-			if resp.GetCode() != 202 {
+			if code != 202 {
 				level.Error(logger).Log("msg", "push error", "err", err)
 			}
-			http.Error(w, string(resp.Body), int(resp.Code))
+			http.Error(w, msg, code)
 		}
 	})
 }
+
+func getHTTPStatusAndMessage(err error) (int, string) {
+	var distributorError distributor_error.DistributorError
+	if errors.As(err, &distributorError) {
+		msg := distributorError.Error()
+		code := GetDistributorPushErrorHTTPStatus(distributorError)
+		return code, msg
+	}
+	return http.StatusInternalServerError, err.Error()
+}
+
+func GetDistributorPushErrorHTTPStatus(distributorError distributor_error.DistributorError) int {
```

Comment:
Shouldn't be exported, and should't have the `get` prefix.

```suggestion
func distributorPushErrorHTTPStatus(distributorError distributor_error.DistributorError) int {
```